File: tests/python/test_chande.py
"""
Python binding tests for Chande indicator.
These tests mirror the Rust unit tests to ensure Python bindings work correctly.
"""
import pytest
import numpy as np
import sys
from pathlib import Path

try:
    import my_project as ta_indicators
except ImportError:

    try:
        import my_project as ta_indicators
    except ImportError:
        pytest.skip("Python module not built. Run 'maturin develop --features python' first", allow_module_level=True)

from test_utils import load_test_data, assert_close, EXPECTED_OUTPUTS
from rust_comparison import compare_with_rust


class TestChande:
    @pytest.fixture(scope='class')
    def test_data(self):
        return load_test_data()

    def test_chande_partial_params(self, test_data):
        """Test Chande with partial parameters - mirrors check_chande_partial_params"""
        high = test_data['high']
        low = test_data['low']
        close = test_data['close']


        result = ta_indicators.chande(high, low, close, 22, 3.0, 'long')
        assert len(result) == len(close)

    def test_chande_accuracy(self, test_data):
        """Test Chande matches expected values from Rust tests - mirrors check_chande_accuracy"""
        high = test_data['high']
        low = test_data['low']
        close = test_data['close']
        expected = EXPECTED_OUTPUTS['chande']

        result = ta_indicators.chande(
            high, low, close,
            period=expected['default_params']['period'],
            mult=expected['default_params']['mult'],
            direction=expected['default_params']['direction']
        )

        assert len(result) == len(close)


        assert_close(
            result[-5:],
            expected['last_5_values'],
            rtol=1e-8,
            msg="Chande last 5 values mismatch"
        )


        warmup_period = expected['warmup_period']
        assert np.all(np.isnan(result[:warmup_period])), f"Expected NaN in warmup period (first {warmup_period} values)"
        assert not np.isnan(result[warmup_period]), f"Expected valid value at index {warmup_period} (after warmup)"


        compare_with_rust('chande', result, 'candles', expected['default_params'])

    def test_chande_zero_period(self):
        """Test Chande fails with zero period - mirrors check_chande_zero_period"""
        high = np.array([10.0, 20.0, 30.0])
        low = np.array([5.0, 15.0, 25.0])
        close = np.array([8.0, 18.0, 28.0])

        with pytest.raises(ValueError, match="Invalid period"):
            ta_indicators.chande(high, low, close, period=0, mult=3.0, direction='long')

    def test_chande_period_exceeds_length(self):
        """Test Chande fails when period exceeds data length - mirrors check_chande_period_exceeds_length"""
        high = np.array([10.0, 20.0, 30.0])
        low = np.array([5.0, 15.0, 25.0])
        close = np.array([8.0, 18.0, 28.0])

        with pytest.raises(ValueError, match="Invalid period"):
            ta_indicators.chande(high, low, close, period=10, mult=3.0, direction='long')

    def test_chande_bad_direction(self):
        """Test Chande fails with bad direction - mirrors check_chande_bad_direction"""
        high = np.array([10.0, 20.0, 30.0, 40.0, 50.0])
        low = np.array([5.0, 15.0, 25.0, 35.0, 45.0])
        close = np.array([8.0, 18.0, 28.0, 38.0, 48.0])

        with pytest.raises(ValueError, match="Invalid direction"):
            ta_indicators.chande(high, low, close, period=2, mult=3.0, direction='bad')

    def test_chande_empty_input(self):
        """Test Chande fails with empty input"""
        empty = np.array([])

        with pytest.raises(ValueError, match="Input series are empty"):
            ta_indicators.chande(empty, empty, empty, period=22, mult=3.0, direction='long')

    def test_chande_all_nan(self):
        """Test Chande fails with all NaN input"""
        all_nan = np.full(100, np.nan)

        with pytest.raises(ValueError, match="All values are NaN"):
            ta_indicators.chande(all_nan, all_nan, all_nan, period=22, mult=3.0, direction='long')

    def test_chande_mismatched_lengths(self):
        """Test Chande fails with mismatched input lengths"""
        high = np.array([10.0, 20.0, 30.0])
        low = np.array([5.0, 15.0])
        close = np.array([8.0, 18.0, 28.0])

        with pytest.raises(ValueError, match="length mismatch"):
            ta_indicators.chande(high, low, close, period=2, mult=3.0, direction='long')

    def test_chande_directions(self):
        """Test Chande with different directions"""
        high = np.array([10.0, 20.0, 30.0, 40.0, 50.0])
        low = np.array([5.0, 15.0, 25.0, 35.0, 45.0])
        close = np.array([8.0, 18.0, 28.0, 38.0, 48.0])


        result_long = ta_indicators.chande(high, low, close, period=3, mult=2.0, direction='long')
        assert len(result_long) == len(close)
        assert np.all(np.isnan(result_long[:2]))


        result_short = ta_indicators.chande(high, low, close, period=3, mult=2.0, direction='short')
        assert len(result_short) == len(close)
        assert np.all(np.isnan(result_short[:2]))


        assert not np.allclose(result_long[2:], result_short[2:], equal_nan=True)

    def test_chande_nan_handling(self, test_data):
        """Test Chande handles NaN values correctly - mirrors check_chande_nan_handling"""
        high = test_data['high']
        low = test_data['low']
        close = test_data['close']
        expected = EXPECTED_OUTPUTS['chande']

        result = ta_indicators.chande(
            high, low, close,
            period=expected['default_params']['period'],
            mult=expected['default_params']['mult'],
            direction=expected['default_params']['direction']
        )
        assert len(result) == len(close)


        warmup_period = expected['warmup_period']
        assert np.all(np.isnan(result[:warmup_period])), f"Expected NaN in warmup period (first {warmup_period} values)"


        if len(result) > warmup_period:
            assert not np.any(np.isnan(result[warmup_period:])), f"Found unexpected NaN after warmup period (index {warmup_period}+)"

    def test_chande_streaming(self, test_data):
        """Test Chande streaming functionality - mirrors check_chande_streaming"""
        high = test_data['high']
        low = test_data['low']
        close = test_data['close']


        batch_result = ta_indicators.chande(high, low, close, period=22, mult=3.0, direction='long')


        stream = ta_indicators.ChandeStream(period=22, mult=3.0, direction='long')
        stream_result = []

        for i in range(len(close)):
            value = stream.update(high[i], low[i], close[i])
            stream_result.append(value if value is not None else np.nan)

        stream_result = np.array(stream_result)


        assert_close(
            batch_result,
            stream_result,
            rtol=1e-8,
            msg="Chande streaming vs batch mismatch"
        )

    def test_chande_batch_single_params(self, test_data):
        """Test Chande batch with single parameter set"""
        high = test_data['high'][:100]
        low = test_data['low'][:100]
        close = test_data['close'][:100]


        result = ta_indicators.chande_batch(
            high, low, close,
            period_range=(22, 22, 0),
            mult_range=(3.0, 3.0, 0.0),
            direction='long'
        )


        assert result['values'].shape[0] == 1
        assert result['values'].shape[1] == 100


        single_result = ta_indicators.chande(high, low, close, 22, 3.0, 'long')
        assert_close(
            result['values'][0],
            single_result,
            rtol=1e-10,
            msg="Batch vs single mismatch"
        )


        assert len(result['periods']) == 1
        assert result['periods'][0] == 22
        assert len(result['mults']) == 1
        assert result['mults'][0] == 3.0
        assert len(result['directions']) == 1
        assert result['directions'][0] == 'long'

    def test_chande_batch_multiple_params(self, test_data):
        """Test Chande batch with multiple parameter combinations"""
        high = test_data['high'][:50]
        low = test_data['low'][:50]
        close = test_data['close'][:50]


        result = ta_indicators.chande_batch(
            high, low, close,
            period_range=(10, 20, 10),
            mult_range=(2.0, 3.0, 0.5),
            direction='short'
        )


        assert result['values'].shape[0] == 6
        assert result['values'].shape[1] == 50


        assert len(result['periods']) == 6
        assert len(result['mults']) == 6
        assert len(result['directions']) == 6


        assert all(d == 'short' for d in result['directions'])


        expected_params = [
            (10, 2.0), (10, 2.5), (10, 3.0),
            (20, 2.0), (20, 2.5), (20, 3.0)
        ]

        for i, (period, mult) in enumerate(expected_params):
            single_result = ta_indicators.chande(high, low, close, period, mult, 'short')
            assert_close(
                result['values'][i],
                single_result,
                rtol=1e-10,
                msg=f"Batch row {i} (period={period}, mult={mult}) mismatch"
            )

    def test_chande_batch_metadata(self, test_data):
        """Test Chande batch result metadata structure"""
        high = test_data['high'][:50]
        low = test_data['low'][:50]
        close = test_data['close'][:50]


        result = ta_indicators.chande_batch(
            high, low, close,
            period_range=(10, 20, 5),
            mult_range=(2.0, 3.0, 0.5),
            direction='long'
        )


        assert 'values' in result
        assert 'periods' in result
        assert 'mults' in result
        assert 'directions' in result


        expected_combos = 9
        assert result['values'].shape[0] == expected_combos
        assert len(result['periods']) == expected_combos
        assert len(result['mults']) == expected_combos
        assert len(result['directions']) == expected_combos


        expected_periods = [10, 10, 10, 15, 15, 15, 20, 20, 20]
        expected_mults = [2.0, 2.5, 3.0] * 3

        assert_close(result['periods'], expected_periods, rtol=0, msg="Period combinations mismatch")
        assert_close(result['mults'], expected_mults, rtol=1e-10, msg="Mult combinations mismatch")
        assert all(d == 'long' for d in result['directions'])

    def test_chande_batch_edge_cases(self, test_data):
        """Test Chande batch with edge cases"""
        high = test_data['high'][:30]
        low = test_data['low'][:30]
        close = test_data['close'][:30]


        result = ta_indicators.chande_batch(
            high, low, close,
            period_range=(10, 12, 5),
            mult_range=(2.0, 2.0, 0.0),
            direction='short'
        )

        assert result['values'].shape[0] == 1
        assert result['periods'][0] == 10
        assert result['mults'][0] == 2.0
        assert result['directions'][0] == 'short'


        result2 = ta_indicators.chande_batch(
            high, low, close,
            period_range=(10, 15, 5),
            mult_range=(2.0, 2.5, 0.5),
            direction='long'
        )

        assert result2['values'].shape[0] == 4
        assert len(result2['periods']) == 4

    def test_chande_warmup_validation(self, test_data):
        """Test Chande warmup period is correctly calculated"""
        high = test_data['high'][:100]
        low = test_data['low'][:100]
        close = test_data['close'][:100]

        test_periods = [5, 10, 22, 50]

        for period in test_periods:
            result = ta_indicators.chande(high, low, close, period=period, mult=2.0, direction='long')

            expected_warmup = period - 1


            for i in range(expected_warmup):
                assert np.isnan(result[i]), f"Expected NaN at index {i} for period {period}"


            if expected_warmup < len(result):
                assert not np.isnan(result[expected_warmup]), f"Expected valid value at index {expected_warmup} for period {period}"

    def test_chande_kernel_parameter(self, test_data):
        """Test Chande with different kernel parameters"""
        high = test_data['high'][:100]
        low = test_data['low'][:100]
        close = test_data['close'][:100]


        kernels = ['auto', 'scalar', 'avx2', 'avx512']
        results = []

        for kernel in kernels:
            try:
                result = ta_indicators.chande(
                    high, low, close,
                    period=22, mult=3.0, direction='long',
                    kernel=kernel
                )
                results.append((kernel, result))
            except ValueError as e:


                msg = str(e).lower()
                allowed = (
                    "invalid kernel" in msg
                    or "not compiled" in msg
                    or "unsupported" in msg
                    or "unavailable" in msg
                    or "not available" in msg
                )
                if not allowed:
                    raise


        if len(results) > 1:
            base_kernel, base_result = results[0]
            for kernel, result in results[1:]:
                assert_close(
                    result,
                    base_result,
                    rtol=1e-10,
                    msg=f"Kernel {kernel} vs {base_kernel} mismatch"
                )


if __name__ == "__main__":
    pytest.main([__file__])

File: tests/wasm/test_chande.js

import test from 'node:test';
import assert from 'node:assert';
import path from 'path';
import { fileURLToPath } from 'url';
import {
    loadTestData,
    assertArrayClose,
    assertClose,
    isNaN,
    assertAllNaN,
    assertNoNaN,
    EXPECTED_OUTPUTS
} from './test_utils.js';
import { compareWithRust } from './rust-comparison.js';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

let wasm;
let testData;

test.before(async () => {

    try {
        const wasmPath = path.join(__dirname, '../../pkg/vector_ta.js');
        const importPath = process.platform === 'win32'
            ? 'file:///' + wasmPath.replace(/\\/g, '/')
            : wasmPath;
        wasm = await import(importPath);

    } catch (error) {
        console.error('Failed to load WASM module. Run "wasm-pack build --features wasm --target nodejs" first');
        throw error;
    }

    testData = loadTestData();
});

test('Chande partial params', () => {

    const high = new Float64Array(testData.high);
    const low = new Float64Array(testData.low);
    const close = new Float64Array(testData.close);

    const result = wasm.chande_js(high, low, close, 22, 3.0, 'long');
    assert.strictEqual(result.length, close.length);
});

test('Chande accuracy', async () => {

    const high = new Float64Array(testData.high);
    const low = new Float64Array(testData.low);
    const close = new Float64Array(testData.close);
    const expected = EXPECTED_OUTPUTS.chande;

    const result = wasm.chande_js(
        high, low, close,
        expected.defaultParams.period,
        expected.defaultParams.mult,
        expected.defaultParams.direction
    );

    assert.strictEqual(result.length, close.length);


    const last5 = result.slice(-5);
    assertArrayClose(
        last5,
        expected.last5Values,
        1e-8,
        "Chande last 5 values mismatch"
    );


    const warmupPeriod = expected.warmupPeriod;
    for (let i = 0; i < warmupPeriod; i++) {
        assert(isNaN(result[i]), `Expected NaN at index ${i} during warmup period`);
    }

    assert(!isNaN(result[warmupPeriod]), `Expected valid value at index ${warmupPeriod} (after warmup)`);


    await compareWithRust('chande', result, 'candles', expected.defaultParams);
});

test('Chande zero period', () => {

    const high = new Float64Array([10.0, 20.0, 30.0]);
    const low = new Float64Array([5.0, 15.0, 25.0]);
    const close = new Float64Array([8.0, 18.0, 28.0]);

    assert.throws(() => {
        wasm.chande_js(high, low, close, 0, 3.0, 'long');
    }, /invalid period/i);
});

test('Chande period exceeds length', () => {

    const high = new Float64Array([10.0, 20.0, 30.0]);
    const low = new Float64Array([5.0, 15.0, 25.0]);
    const close = new Float64Array([8.0, 18.0, 28.0]);

    assert.throws(() => {
        wasm.chande_js(high, low, close, 10, 3.0, 'long');
    }, /invalid period/i);
});

test('Chande bad direction', () => {

    const high = new Float64Array([10.0, 20.0, 30.0, 40.0, 50.0]);
    const low = new Float64Array([5.0, 15.0, 25.0, 35.0, 45.0]);
    const close = new Float64Array([8.0, 18.0, 28.0, 38.0, 48.0]);

    assert.throws(() => {
        wasm.chande_js(high, low, close, 2, 3.0, 'bad');
    }, /invalid direction/i);
});

test('Chande empty input', () => {

    const empty = new Float64Array([]);

    assert.throws(() => {
        wasm.chande_js(empty, empty, empty, 22, 3.0, 'long');
    }, /input series are empty/i);
});

test('Chande mismatched lengths', () => {

    const high = new Float64Array([10.0, 20.0, 30.0]);
    const low = new Float64Array([5.0, 15.0]);
    const close = new Float64Array([8.0, 18.0, 28.0]);

    assert.throws(() => {
        wasm.chande_js(high, low, close, 2, 3.0, 'long');
    }, /length mismatch/);
});

test('Chande directions', () => {

    const high = new Float64Array([10.0, 20.0, 30.0, 40.0, 50.0]);
    const low = new Float64Array([5.0, 15.0, 25.0, 35.0, 45.0]);
    const close = new Float64Array([8.0, 18.0, 28.0, 38.0, 48.0]);


    const resultLong = wasm.chande_js(high, low, close, 3, 2.0, 'long');
    assert.strictEqual(resultLong.length, close.length);
    assert(isNaN(resultLong[0]));
    assert(isNaN(resultLong[1]));
    assert(!isNaN(resultLong[2]));


    const resultShort = wasm.chande_js(high, low, close, 3, 2.0, 'short');
    assert.strictEqual(resultShort.length, close.length);
    assert(isNaN(resultShort[0]));
    assert(isNaN(resultShort[1]));
    assert(!isNaN(resultShort[2]));


    assert.notStrictEqual(resultLong[2], resultShort[2]);
});

test('Chande batch single params', () => {

    const high = new Float64Array(testData.high.slice(0, 100));
    const low = new Float64Array(testData.low.slice(0, 100));
    const close = new Float64Array(testData.close.slice(0, 100));


    const result = wasm.chande_batch_js(
        high, low, close,
        22, 22, 0,
        3.0, 3.0, 0.0,
        'long'
    );


    assert.strictEqual(result.rows, 1);
    assert.strictEqual(result.cols, 100);
    assert.strictEqual(result.values.length, 100);


    const singleResult = wasm.chande_js(high, low, close, 22, 3.0, 'long');
    assertArrayClose(
        result.values,
        singleResult,
        1e-10,
        "Batch vs single mismatch"
    );


    assert.strictEqual(result.periods.length, 1);
    assert.strictEqual(result.periods[0], 22);
    assert.strictEqual(result.mults.length, 1);
    assert.strictEqual(result.mults[0], 3.0);
    assert.strictEqual(result.directions.length, 1);
    assert.strictEqual(result.directions[0], 'long');
});

test('Chande batch multiple params', () => {

    const high = new Float64Array(testData.high.slice(0, 50));
    const low = new Float64Array(testData.low.slice(0, 50));
    const close = new Float64Array(testData.close.slice(0, 50));


    const result = wasm.chande_batch_js(
        high, low, close,
        10, 20, 10,
        2.0, 3.0, 0.5,
        'short'
    );


    assert.strictEqual(result.rows, 6);
    assert.strictEqual(result.cols, 50);
    assert.strictEqual(result.values.length, 6 * 50);


    assert.strictEqual(result.periods.length, 6);
    assert.strictEqual(result.mults.length, 6);
    assert.strictEqual(result.directions.length, 6);


    assert(result.directions.every(d => d === 'short'));


    const expectedParams = [
        [10, 2.0], [10, 2.5], [10, 3.0],
        [20, 2.0], [20, 2.5], [20, 3.0]
    ];

    for (let i = 0; i < expectedParams.length; i++) {
        const [period, mult] = expectedParams[i];
        const singleResult = wasm.chande_js(high, low, close, period, mult, 'short');
        const batchRow = result.values.slice(i * 50, (i + 1) * 50);

        assertArrayClose(
            batchRow,
            singleResult,
            1e-10,
            `Batch row ${i} (period=${period}, mult=${mult}) mismatch`
        );
    }
});

test('Chande zero-copy API', () => {

    const len = 100;
    const high = new Float64Array(testData.high.slice(0, len));
    const low = new Float64Array(testData.low.slice(0, len));
    const close = new Float64Array(testData.close.slice(0, len));


    const highPtr = wasm.chande_alloc(len);
    const lowPtr = wasm.chande_alloc(len);
    const closePtr = wasm.chande_alloc(len);
    const outPtr = wasm.chande_alloc(len);

    try {

        const highMem = new Float64Array(wasm.__wasm.memory.buffer, highPtr, len);
        const lowMem = new Float64Array(wasm.__wasm.memory.buffer, lowPtr, len);
        const closeMem = new Float64Array(wasm.__wasm.memory.buffer, closePtr, len);
        highMem.set(high);
        lowMem.set(low);
        closeMem.set(close);


        wasm.chande_into(
            highPtr,
            lowPtr,
            closePtr,
            outPtr,
            len,
            22,
            3.0,
            'long'
        );


        const result = new Float64Array(wasm.__wasm.memory.buffer, outPtr, len);


        const expected = wasm.chande_js(high, low, close, 22, 3.0, 'long');
        assertArrayClose(
            result,
            expected,
            1e-10,
            "Zero-copy API mismatch"
        );
    } finally {

        wasm.chande_free(highPtr, len);
        wasm.chande_free(lowPtr, len);
        wasm.chande_free(closePtr, len);
        wasm.chande_free(outPtr, len);
    }
});

test('Chande zero-copy API with aliasing', () => {

    const len = 100;
    const high = new Float64Array(testData.high.slice(0, len));
    const low = new Float64Array(testData.low.slice(0, len));
    const close = new Float64Array(testData.close.slice(0, len));


    const highPtr = wasm.chande_alloc(len);
    const lowPtr = wasm.chande_alloc(len);
    const closePtr = wasm.chande_alloc(len);

    try {

        const highMem = new Float64Array(wasm.__wasm.memory.buffer, highPtr, len);
        const lowMem = new Float64Array(wasm.__wasm.memory.buffer, lowPtr, len);
        const closeMem = new Float64Array(wasm.__wasm.memory.buffer, closePtr, len);
        highMem.set(high);
        lowMem.set(low);
        closeMem.set(close);


        wasm.chande_into(
            highPtr,
            lowPtr,
            closePtr,
            closePtr,
            len,
            22,
            3.0,
            'long'
        );


        const result = new Float64Array(wasm.__wasm.memory.buffer, closePtr, len);


        const expected = wasm.chande_js(high, low, close, 22, 3.0, 'long');
        assertArrayClose(
            result,
            expected,
            1e-10,
            "Zero-copy API with aliasing mismatch"
        );
    } finally {

        wasm.chande_free(highPtr, len);
        wasm.chande_free(lowPtr, len);
        wasm.chande_free(closePtr, len);
    }
});


test('Chande batch metadata structure', () => {

    const high = new Float64Array(testData.high.slice(0, 50));
    const low = new Float64Array(testData.low.slice(0, 50));
    const close = new Float64Array(testData.close.slice(0, 50));
    const expected = EXPECTED_OUTPUTS.chande;


    const result = wasm.chande_batch_js(
        high, low, close,
        expected.defaultParams.period, expected.defaultParams.period, 0,
        expected.defaultParams.mult, expected.defaultParams.mult, 0,
        expected.defaultParams.direction
    );


    assert(result.values, 'Should have values array');
    assert(result.periods, 'Should have periods array');
    assert(result.mults, 'Should have mults array');
    assert(result.directions, 'Should have directions array');
    assert(typeof result.rows === 'number', 'Should have rows count');
    assert(typeof result.cols === 'number', 'Should have cols count');


    assert.strictEqual(result.rows, 1);
    assert.strictEqual(result.cols, 50);
    assert.strictEqual(result.periods.length, 1);
    assert.strictEqual(result.values.length, 50);
});

test('Chande batch multiple parameter combinations', () => {

    const high = new Float64Array(testData.high.slice(0, 50));
    const low = new Float64Array(testData.low.slice(0, 50));
    const close = new Float64Array(testData.close.slice(0, 50));

    const result = wasm.chande_batch_js(
        high, low, close,
        10, 20, 10,
        2.0, 3.0, 0.5,
        'short'
    );


    assert.strictEqual(result.rows, 6);
    assert.strictEqual(result.cols, 50);
    assert.strictEqual(result.periods.length, 6);
    assert.strictEqual(result.mults.length, 6);
    assert.strictEqual(result.directions.length, 6);
    assert.strictEqual(result.values.length, 300);


    const expectedPeriods = [10, 10, 10, 20, 20, 20];
    const expectedMults = [2.0, 2.5, 3.0, 2.0, 2.5, 3.0];

    for (let i = 0; i < 6; i++) {
        assert.strictEqual(result.periods[i], expectedPeriods[i], `Period mismatch at index ${i}`);
        assertClose(result.mults[i], expectedMults[i], 1e-10, `Mult mismatch at index ${i}`);
        assert.strictEqual(result.directions[i], 'short', `Direction mismatch at index ${i}`);
    }


    const firstRow = result.values.slice(0, result.cols);
    const singleResult = wasm.chande_js(high, low, close, 10, 2.0, 'short');
    assertArrayClose(firstRow, singleResult, 1e-10, "First batch row vs single mismatch");
});

test('Chande batch edge cases', () => {

    const high = new Float64Array([10, 20, 30, 40, 50, 60, 70, 80, 90, 100]);
    const low = new Float64Array([5, 15, 25, 35, 45, 55, 65, 75, 85, 95]);
    const close = new Float64Array([8, 18, 28, 38, 48, 58, 68, 78, 88, 98]);


    const singleBatch = wasm.chande_batch_js(
        high, low, close,
        5, 5, 0,
        2.0, 2.0, 0,
        'long'
    );

    assert.strictEqual(singleBatch.rows, 1);
    assert.strictEqual(singleBatch.values.length, 10);
    assert.strictEqual(singleBatch.periods.length, 1);


    const largeBatch = wasm.chande_batch_js(
        high, low, close,
        5, 7, 10,
        2.0, 2.0, 0,
        'long'
    );


    assert.strictEqual(largeBatch.rows, 1);
    assert.strictEqual(largeBatch.values.length, 10);
    assert.strictEqual(largeBatch.periods[0], 5);
});

test('Chande warmup period validation', () => {

    const high = new Float64Array(100).fill(100);
    const low = new Float64Array(100).fill(90);
    const close = new Float64Array(100).fill(95);

    const testPeriods = [5, 10, 22, 30];

    for (const period of testPeriods) {
        const result = wasm.chande_js(high, low, close, period, 2.0, 'long');
        const expectedWarmup = period - 1;


        for (let i = 0; i < expectedWarmup; i++) {
            assert(isNaN(result[i]), `Expected NaN at index ${i} for period ${period}`);
        }


        assert(!isNaN(result[expectedWarmup]), `Expected valid value at index ${expectedWarmup} for period ${period}`);
    }
});

test('Chande SIMD128 consistency', () => {


    const testCases = [
        { size: 10, period: 5 },
        { size: 100, period: 22 },
        { size: 1000, period: 50 },
        { size: 10000, period: 100 }
    ];

    for (const testCase of testCases) {
        const high = new Float64Array(testCase.size);
        const low = new Float64Array(testCase.size);
        const close = new Float64Array(testCase.size);

        for (let i = 0; i < testCase.size; i++) {
            const base = 100 + Math.sin(i * 0.1) * 10;
            high[i] = base + 5;
            low[i] = base - 5;
            close[i] = base + Math.cos(i * 0.05) * 3;
        }

        const result = wasm.chande_js(high, low, close, testCase.period, 3.0, 'long');


        assert.strictEqual(result.length, testCase.size);


        const expectedWarmup = testCase.period - 1;
        for (let i = 0; i < expectedWarmup; i++) {
            assert(isNaN(result[i]), `Expected NaN at warmup index ${i} for size=${testCase.size}`);
        }


        for (let i = expectedWarmup; i < Math.min(expectedWarmup + 10, result.length); i++) {
            assert(!isNaN(result[i]), `Unexpected NaN at index ${i} for size=${testCase.size}`);
        }
    }
});

test('Chande zero-copy error handling', () => {

    assert.throws(() => {
        wasm.chande_into(0, 0, 0, 0, 10, 22, 3.0, 'long');
    }, /null pointer|invalid memory/i);


    const highPtr = wasm.chande_alloc(10);
    const lowPtr = wasm.chande_alloc(10);
    const closePtr = wasm.chande_alloc(10);
    const outPtr = wasm.chande_alloc(10);

    try {

        assert.throws(() => {
            wasm.chande_into(highPtr, lowPtr, closePtr, outPtr, 10, 0, 3.0, 'long');
        }, /invalid period/i);


        assert.throws(() => {
            wasm.chande_into(highPtr, lowPtr, closePtr, outPtr, 10, 5, 3.0, 'invalid');
        }, /invalid direction/i);
    } finally {
        wasm.chande_free(highPtr, 10);
        wasm.chande_free(lowPtr, 10);
        wasm.chande_free(closePtr, 10);
        wasm.chande_free(outPtr, 10);
    }
});

test('Chande memory management', () => {

    const sizes = [100, 1000, 10000];

    for (const size of sizes) {
        const highPtr = wasm.chande_alloc(size);
        const lowPtr = wasm.chande_alloc(size);
        const closePtr = wasm.chande_alloc(size);
        const outPtr = wasm.chande_alloc(size);

        assert(highPtr !== 0, `Failed to allocate high buffer for ${size} elements`);
        assert(lowPtr !== 0, `Failed to allocate low buffer for ${size} elements`);
        assert(closePtr !== 0, `Failed to allocate close buffer for ${size} elements`);
        assert(outPtr !== 0, `Failed to allocate output buffer for ${size} elements`);


        const highMem = new Float64Array(wasm.__wasm.memory.buffer, highPtr, size);
        const lowMem = new Float64Array(wasm.__wasm.memory.buffer, lowPtr, size);
        const closeMem = new Float64Array(wasm.__wasm.memory.buffer, closePtr, size);

        for (let i = 0; i < Math.min(10, size); i++) {
            highMem[i] = 100 + i;
            lowMem[i] = 90 + i;
            closeMem[i] = 95 + i;
        }


        for (let i = 0; i < Math.min(10, size); i++) {
            assert.strictEqual(highMem[i], 100 + i, `High memory corruption at index ${i}`);
            assert.strictEqual(lowMem[i], 90 + i, `Low memory corruption at index ${i}`);
            assert.strictEqual(closeMem[i], 95 + i, `Close memory corruption at index ${i}`);
        }


        wasm.chande_free(highPtr, size);
        wasm.chande_free(lowPtr, size);
        wasm.chande_free(closePtr, size);
        wasm.chande_free(outPtr, size);
    }
});

test('Chande batch zero-copy API', () => {

    const len = 50;
    const high = new Float64Array(testData.high.slice(0, len));
    const low = new Float64Array(testData.low.slice(0, len));
    const close = new Float64Array(testData.close.slice(0, len));


    const periodStart = 10, periodEnd = 20, periodStep = 10;
    const multStart = 2.0, multEnd = 3.0, multStep = 0.5;


    const expectedRows = 2 * 3;


    const highPtr = wasm.chande_alloc(len);
    const lowPtr = wasm.chande_alloc(len);
    const closePtr = wasm.chande_alloc(len);
    const outPtr = wasm.chande_alloc(expectedRows * len);

    try {

        const highMem = new Float64Array(wasm.__wasm.memory.buffer, highPtr, len);
        const lowMem = new Float64Array(wasm.__wasm.memory.buffer, lowPtr, len);
        const closeMem = new Float64Array(wasm.__wasm.memory.buffer, closePtr, len);
        highMem.set(high);
        lowMem.set(low);
        closeMem.set(close);


        const rows = wasm.chande_batch_into(
            highPtr,
            lowPtr,
            closePtr,
            outPtr,
            len,
            periodStart, periodEnd, periodStep,
            multStart, multEnd, multStep,
            'long'
        );

        assert.strictEqual(rows, expectedRows);


        const result = new Float64Array(wasm.__wasm.memory.buffer, outPtr, rows * len);


        const expected = wasm.chande_batch_js(
            high, low, close,
            periodStart, periodEnd, periodStep,
            multStart, multEnd, multStep,
            'long'
        );

        assertArrayClose(
            result,
            expected.values,
            1e-10,
            "Batch zero-copy API mismatch"
        );
    } finally {

        wasm.chande_free(highPtr, len);
        wasm.chande_free(lowPtr, len);
        wasm.chande_free(closePtr, len);
        wasm.chande_free(outPtr, expectedRows * len);
    }
});
